In [38]:
#import Libraries
import pandas as pd
import numpy as np
import os

1.Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

Done

In [39]:
path=r"C:\Users\abdul\Career-Foundry-Assignment4-Instacart-Basket-Analysis"

In [40]:
ords_prods_merge=pd.read_pickle(os.path.join(path, '02_Data','Prepared-Data','orders_products_merged_updated.pkl' ))

In [41]:
ords_prods_merge.head()

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0             0             0   2539329        1    prior             1   
1             1             1   2398795        1    prior             2   
2             2             2    473747        1    prior             3   
3             3             3   2254736        1    prior             4   
4             4             4    431534        1    prior             5   

   order_dow  order_hour_of_day  days_since_prior_order  product_id  \
0          2                  8                     7.0         196   
1          3                  7                    15.0         196   
2          3                 12                    21.0         196   
3          4                  7                    29.0         196   
4          4                 15                    28.0         196   

   add_to_cart_order  reordered  Unnamed: 0_y product_name  aisle_id  \
0                  1          0           195         Soda        77   
1                  1          1           195         Soda        77   
2                  1          1           195         Soda        77   
3                  1          1           195         Soda        77   
4                  1          1           195         Soda        77   

   department_id  prices         Busiest days busiest_period_of_day  
0              7     9.0  Regularly Busy Days        Average Orders  
1              7     9.0      Least Busy Days        Average Orders  
2              7     9.0      Least Busy Days           Most Orders  
3              7     9.0      Least Busy Days        Average Orders  
4              7     9.0      Least Busy Days           Most Orders

2.In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [42]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

3.Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

Answer: In the Subset there are only 8 entries but in the entire dataframe there are 21 entries.

4.Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

Answer: Done

5.The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [43]:
ords_prods_merge['max_order']=ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [44]:
ords_prods_merge.head(60)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0              0             0   2539329        1    prior             1   
1              1             1   2398795        1    prior             2   
2              2             2    473747        1    prior             3   
3              3             3   2254736        1    prior             4   
4              4             4    431534        1    prior             5   
5              5             5   3367565        1    prior             6   
6              6             6    550135        1    prior             7   
7              7             7   3108588        1    prior             8   
8              8             8   2295261        1    prior             9   
9              9             9   2550362        1    prior            10   
10           144           144   2968173       15    prior            15   
11           146           146   1870022       15    prior            17   
12           147           147   1911383       15    prior            18   
13           150           150   2715276       15    prior            21   
14           151           151    487368       15    prior            22   
15           209           209   2293453       19    prior             2   
16           212           212   1973799       19    prior             5   
17           214           214    532817       19    prior             7   
18           232           232   1573906       21    prior            10   
19           458           458   1593000       31    prior            10   
20           465           465   2231262       31    prior            17   
21           617           617   2580647       43    prior             6   
22           620           620   2187180       43    prior             9   
23           751           751   2497897       52    prior             1   
24           752           752   1318871       52    prior             2   
25           753           753   1261384       52    prior             3   
26           755           755    580568       52    prior             5   
27           757           757   2428073       52    prior             7   
28           759           759    342306       52    prior             9   
29           760           760   1520353       52    prior            10   
30           761           761   2542086       52    prior            11   
31           762           762    944694       52    prior            12   
32           765           765   1268191       52    prior            15   
33           767           767   1122089       52    prior            17   
34           768           768   1498922       52    prior            18   
35           769           769    180919       52    prior            19   
36           771           771   2695875       52    prior            21   
37          1022          1022   3205960       67    prior             1   
38          1023          1023    970922       67    prior             2   
39          1024          1024    596585       67    prior             3   
40          1025          1025   2037087       67    prior             4   
41          1026          1026   2474111       67    prior             5   
42          1027          1027   1568481       67    prior             6   
43          1029          1029   2048527       67    prior             8   
44          1030          1030   2389031       67    prior             9   
45          1031          1031   2378562       67    prior            10   
46          1032          1032   1515698       67    prior            11   
47          1033          1033   1928050       67    prior            12   
48          1034          1034   1949274       67    prior            13   
49          1035          1035    551869       67    prior            14   
50          1037          1037    768572       67    prior            16   
51          1039          1039    150528       67   

In [45]:
ords_prods_merge.loc[ords_prods_merge['max_order']>40, 'loyalty_flag']='Loyal Customer'

In [46]:
ords_prods_merge.loc[(ords_prods_merge['max_order']<=40) & (ords_prods_merge['max_order']>10), 'loyalty_flag']='Regular Customer'

In [47]:
ords_prods_merge.loc[ords_prods_merge['max_order']<=10, 'loyalty_flag']='New Customer'

In [48]:
ords_prods_merge.head(15)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0              0             0   2539329        1    prior             1   
1              1             1   2398795        1    prior             2   
2              2             2    473747        1    prior             3   
3              3             3   2254736        1    prior             4   
4              4             4    431534        1    prior             5   
5              5             5   3367565        1    prior             6   
6              6             6    550135        1    prior             7   
7              7             7   3108588        1    prior             8   
8              8             8   2295261        1    prior             9   
9              9             9   2550362        1    prior            10   
10           144           144   2968173       15    prior            15   
11           146           146   1870022       15    prior            17   
12           147           147   1911383       15    prior            18   
13           150           150   2715276       15    prior            21   
14           151           151    487368       15    prior            22   

    order_dow  order_hour_of_day  days_since_prior_order  product_id  ...  \
0           2                  8                     7.0         196  ...   
1           3                  7                    15.0         196  ...   
2           3                 12                    21.0         196  ...   
3           4                  7                    29.0         196  ...   
4           4                 15                    28.0         196  ...   
5           2                  7                    19.0         196  ...   
6           1                  9                    20.0         196  ...   
7           1                 14                    14.0         196  ...   
8           1                 16                     0.0         196  ...   
9           4                  8                    30.0         196  ...   
10          1                  9                     7.0         196  ...   
11          2                 16                     8.0         196  ...   
12          2                 11                     7.0         196  ...   
13          1                  9                     7.0         196  ...   
14          1                 10                    14.0         196  ...   

    reordered  Unnamed: 0_y  product_name aisle_id  department_id  prices  \
0           0           195          Soda       77              7     9.0   
1           1           195          Soda       77              7     9.0   
2           1           195          Soda       77              7     9.0   
3           1           195          Soda       77              7     9.0   
4           1           195          Soda       77              7     9.0   
5           1           195          Soda       77              7     9.0   
6           1           195          Soda       77              7     9.0   
7           1           195          Soda       77              7     9.0   
8           1           195          Soda       77              7     9.0   
9           1           195          Soda       77              7     9.0   
10          0           195          Soda       77              7     9.0   
11          1           195          Soda       77              7     9.0   
12          1           195          Soda       77              7     9.0   
13          1           195          Soda       77              7     9.0   
14          1           195          Soda       77              7     9.0   

           Busiest days busiest_period_of_day max_order      loyalty_flag  
0   Regularly Busy Days        Average Orders        10      New Customer  
1       Least Busy Days        Average Orders        10      New Customer  
2       Least Busy Days           Most Orders        10      New Customer  
3       Least Bus

In [49]:
ords_prods_merge['loyalty_flag'].value_counts()

Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: loyalty_flag, dtype: int64

In [50]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min','max','sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal Customer    10.386336  1.0  99999.0  106814042.2
New Customer      13.294670  1.0  99999.0   83011787.2
Regular Customer  12.495717  1.0  99999.0  198391693.2

Answer: Loyal coustomers are buying cheaper products when compared to the New and Regular customers. They are giving second highest revenue to the company. 
Regular customers are the highest revenue giving category.Their average spending is 12.49$.

New customers are the highest average spenders with an average of 13.29$.Normally these customers are visiting the site to buy specific products which they can't get in their regular shops.

6.The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [51]:
ords_prods_merge['average_price']=ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [52]:
ords_prods_merge.head(5)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0             0             0   2539329        1    prior             1   
1             1             1   2398795        1    prior             2   
2             2             2    473747        1    prior             3   
3             3             3   2254736        1    prior             4   
4             4             4    431534        1    prior             5   

   order_dow  order_hour_of_day  days_since_prior_order  product_id  ...  \
0          2                  8                     7.0         196  ...   
1          3                  7                    15.0         196  ...   
2          3                 12                    21.0         196  ...   
3          4                  7                    29.0         196  ...   
4          4                 15                    28.0         196  ...   

   Unnamed: 0_y  product_name  aisle_id department_id  prices  \
0           195          Soda        77             7     9.0   
1           195          Soda        77             7     9.0   
2           195          Soda        77             7     9.0   
3           195          Soda        77             7     9.0   
4           195          Soda        77             7     9.0   

          Busiest days  busiest_period_of_day max_order  loyalty_flag  \
0  Regularly Busy Days         Average Orders        10  New Customer   
1      Least Busy Days         Average Orders        10  New Customer   
2      Least Busy Days            Most Orders        10  New Customer   
3      Least Busy Days         Average Orders        10  New Customer   
4      Least Busy Days            Most Orders        10  New Customer   

   average_price  
0       6.367797  
1       6.367797  
2       6.367797  
3       6.367797  
4       6.367797  

[5 rows x 22 columns]

In [53]:
ords_prods_merge.loc[ords_prods_merge['average_price']<10, 'spending_flag']='Low spender'

In [54]:
ords_prods_merge.loc[ords_prods_merge['average_price']>=10, 'spending_flag']='High spender'

In [55]:
ords_prods_merge['spending_flag'].value_counts()

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [56]:
ords_prods_merge.head(10)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0             0             0   2539329        1    prior             1   
1             1             1   2398795        1    prior             2   
2             2             2    473747        1    prior             3   
3             3             3   2254736        1    prior             4   
4             4             4    431534        1    prior             5   
5             5             5   3367565        1    prior             6   
6             6             6    550135        1    prior             7   
7             7             7   3108588        1    prior             8   
8             8             8   2295261        1    prior             9   
9             9             9   2550362        1    prior            10   

   order_dow  order_hour_of_day  days_since_prior_order  product_id  ...  \
0          2                  8                     7.0         196  ...   
1          3                  7                    15.0         196  ...   
2          3                 12                    21.0         196  ...   
3          4                  7                    29.0         196  ...   
4          4                 15                    28.0         196  ...   
5          2                  7                    19.0         196  ...   
6          1                  9                    20.0         196  ...   
7          1                 14                    14.0         196  ...   
8          1                 16                     0.0         196  ...   
9          4                  8                    30.0         196  ...   

   product_name  aisle_id  department_id prices         Busiest days  \
0          Soda        77              7    9.0  Regularly Busy Days   
1          Soda        77              7    9.0      Least Busy Days   
2          Soda        77              7    9.0      Least Busy Days   
3          Soda        77              7    9.0      Least Busy Days   
4          Soda        77              7    9.0      Least Busy Days   
5          Soda        77              7    9.0  Regularly Busy Days   
6          Soda        77              7    9.0         Busiest Days   
7          Soda        77              7    9.0         Busiest Days   
8          Soda        77              7    9.0         Busiest Days   
9          Soda        77              7    9.0      Least Busy Days   

   busiest_period_of_day  max_order  loyalty_flag average_price  spending_flag  
0         Average Orders         10  New Customer      6.367797    Low spender  
1         Average Orders         10  New Customer      6.367797    Low spender  
2            Most Orders         10  New Customer      6.367797    Low spender  
3         Average Orders         10  New Customer      6.367797    Low spender  
4            Most Orders         10  New Customer      6.367797    Low spender  
5         Average Orders         10  New Customer      6.367797    Low spender  
6            Most Orders         10  New Customer      6.367797    Low spender  
7            Most Orders         10  New Customer      6.367797    Low spender  
8            Most Orders         10  New Customer      6.367797    Low spender  
9         Average Orders         10  New Customer      6.367797    Low spender  

[10 rows x 23 columns]

7.In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
-If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
-If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
-If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [57]:
ords_prods_merge['days_since_prior_order_median']=ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [58]:
ords_prods_merge['days_since_prior_order_median'].value_counts()

7.0     7463145
6.0     2787313
8.0     2742855
5.0     2334096
4.0     2202399
         ...   
28.5       4102
27.5       3730
29.5       3022
1.5        1010
0.5          80
Name: days_since_prior_order_median, Length: 61, dtype: int64

In [59]:
ords_prods_merge.loc[ords_prods_merge['days_since_prior_order_median']>20, 'criteria_flag']='Non-frequent customer'

In [60]:
ords_prods_merge.loc[(ords_prods_merge['days_since_prior_order_median']<=20) & (ords_prods_merge['days_since_prior_order_median']>10), 'criteria_flag']='Regular Customer'

In [61]:
ords_prods_merge.loc[ords_prods_merge['days_since_prior_order_median']<=10, 'criteria_flag']='Frequent customer'

In [62]:
ords_prods_merge.head(20)

Unnamed: 0.1  Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0              0             0   2539329        1    prior             1   
1              1             1   2398795        1    prior             2   
2              2             2    473747        1    prior             3   
3              3             3   2254736        1    prior             4   
4              4             4    431534        1    prior             5   
5              5             5   3367565        1    prior             6   
6              6             6    550135        1    prior             7   
7              7             7   3108588        1    prior             8   
8              8             8   2295261        1    prior             9   
9              9             9   2550362        1    prior            10   
10           144           144   2968173       15    prior            15   
11           146           146   1870022       15    prior            17   
12           147           147   1911383       15    prior            18   
13           150           150   2715276       15    prior            21   
14           151           151    487368       15    prior            22   
15           209           209   2293453       19    prior             2   
16           212           212   1973799       19    prior             5   
17           214           214    532817       19    prior             7   
18           232           232   1573906       21    prior            10   
19           458           458   1593000       31    prior            10   

    order_dow  order_hour_of_day  days_since_prior_order  product_id  ...  \
0           2                  8                     7.0         196  ...   
1           3                  7                    15.0         196  ...   
2           3                 12                    21.0         196  ...   
3           4                  7                    29.0         196  ...   
4           4                 15                    28.0         196  ...   
5           2                  7                    19.0         196  ...   
6           1                  9                    20.0         196  ...   
7           1                 14                    14.0         196  ...   
8           1                 16                     0.0         196  ...   
9           4                  8                    30.0         196  ...   
10          1                  9                     7.0         196  ...   
11          2                 16                     8.0         196  ...   
12          2                 11                     7.0         196  ...   
13          1                  9                     7.0         196  ...   
14          1                 10                    14.0         196  ...   
15          5                 14                     6.0         196  ...   
16          6                 12                     8.0         196  ...   
17          4                 17                     6.0         196  ...   
18          3                 10                     6.0         196  ...   
19          3                  8                     7.0         196  ...   

    department_id  prices         Busiest days busiest_period_of_day  \
0               7     9.0  Regularly Busy Days        Average Orders   
1               7     9.0      Least Busy Days        Average Orders   
2               7     9.0      Least Busy Days           Most Orders   
3               7     9.0      Least Busy Days        Average Orders   
4               7     9.0      Least Busy Days           Most Orders   
5               7     9.0  Regularly Busy Days        Average Orders   
6               7     9.0         Busiest Days           Most Orders   
7               7     9.0         Busiest Days           Most Orders   
8               7     9.0         Busiest Days           Most Orders   
9               7     9.0      Least Busy Days        Average Orders 

In [63]:
ords_prods_merge['criteria_flag'].value_counts()

Frequent customer        22790523
Regular Customer          6927608
Non-frequent customer     2686728
Name: criteria_flag, dtype: int64

8.Ensure your notebook is clean and structured and that your code is well commented.

Answer:Done

9.Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [66]:
ords_prods_merge.to_pickle(os.path.join(path,'02_Data','Prepared-Data','ords_prods_merg_updated_48.pkl' ))

10.Save your notebook and submit it to your tutor for review.

Answer:Done